File: kotlin/aksara-core/src/main/kotlin/com/aksara/DiskCache.kt
package com.aksara

import java.io.File

/**
 * Persists the last-good remote bundle (and its ETag) per language.
 *
 * On launch the [Localizer] loads the cached bundle — if any — **before** hitting the
 * network, so a returning user sees the latest translations offline and the first
 * render never waits on a request.
 *
 * On Android, pass `context.cacheDir` (or a subdirectory of it) as [directory].
 * Mirror of the Swift `DiskCache`.
 */
class DiskCache(directory: File? = null) {
    private val directory: File =
        (directory ?: File(System.getProperty("java.io.tmpdir"), "Aksara")).also { it.mkdirs() }

    private fun bundleFile(language: String) = File(directory, "$language.json")
    private fun etagFile(language: String) = File(directory, "$language.etag")

    fun saveBundle(data: ByteArray, etag: String?, language: String) {
        bundleFile(language).writeBytes(data)
        if (etag != null) etagFile(language).writeText(etag)
    }

    fun loadBundle(language: String): ByteArray? =
        bundleFile(language).takeIf { it.exists() }?.readBytes()

    fun loadETag(language: String): String? =
        etagFile(language).takeIf { it.exists() }?.readText()

    fun clear(language: String) {
        bundleFile(language).delete()
        etagFile(language).delete()
    }
}

File: swift/Sources/Aksara/DiskCache.swift
import Foundation

/// Persists the last-good remote bundle (and its ETag) per language.
///
/// On launch the `Localizer` loads the cached bundle — if any — **before** hitting
/// the network, so a returning user sees the latest translations offline and the
/// first render never waits on a request.
public final class DiskCache: @unchecked Sendable {
    private let directory: URL
    private let fileManager = FileManager.default

    /// - Parameter directory: override the cache location (used by tests). Defaults
    ///   to `<Caches>/Aksara`.
    public init(directory: URL? = nil) {
        if let directory {
            self.directory = directory
        } else {
            let base = (try? FileManager.default.url(
                for: .cachesDirectory, in: .userDomainMask, appropriateFor: nil, create: true
            )) ?? FileManager.default.temporaryDirectory
            self.directory = base.appendingPathComponent("Aksara", isDirectory: true)
        }
        try? fileManager.createDirectory(at: self.directory, withIntermediateDirectories: true)
    }

    private func bundleURL(_ language: String) -> URL {
        directory.appendingPathComponent("\(language).json")
    }
    private func etagURL(_ language: String) -> URL {
        directory.appendingPathComponent("\(language).etag")
    }

    public func saveBundle(_ data: Data, etag: String?, language: String) {
        try? data.write(to: bundleURL(language), options: .atomic)
        if let etag {
            try? Data(etag.utf8).write(to: etagURL(language), options: .atomic)
        }
    }

    public func loadBundle(language: String) -> Data? {
        try? Data(contentsOf: bundleURL(language))
    }

    public func loadETag(language: String) -> String? {
        guard let data = try? Data(contentsOf: etagURL(language)) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    public func clear(language: String) {
        try? fileManager.removeItem(at: bundleURL(language))
        try? fileManager.removeItem(at: etagURL(language))
    }
}
